src/backend/distributed/utils/background_jobs.c

```diff
@@ -674,22 +735,49 @@ CitusBackgroundTaskQueueMonitorMain(Datum arg)
 		.ctx = backgroundTaskContext
 	};
 
+	/* flag to prevent duplicate termination of task executors */
+	bool termHandleStarted = false;
+
 	/* loop exits if there is no running tasks left */
 	bool hasAnyTask = true;
 	while (hasAnyTask)
 	{
 		/* handle signals */
 		CHECK_FOR_INTERRUPTS();
 
+		/*
+		 * if the flag is set, we should terminate all task executor workers to prevent duplicate
+		 * runs of the same task on the next start of the monitor, which is dangerous for non-idempotent
+		 * tasks. We do not break the loop here as we want to reflect tasks' messages. Hence, we wait until
+		 * all tasks finish and also do not allow new runnable tasks to start running. After all current tasks
+		 * finish, we can exit the loop safely.
+		 */
+		if (GotSigterm && !termHandleStarted)
+		{
+			ereport(LOG, (errmsg("handling termination signal")));
+			TerminateAllTaskExecutors(backgroundExecutorHandles);
+			termHandleStarted = true;
+		}
+
+		if (GotSighup)
+		{
+			/* update max_background_task_executors if changed */
+			ProcessConfigFile(PGC_SIGHUP);
+			GotSighup = false;
```

Comment:
For me it looks like there is small race condition.
E.g., while the `ProcessConfigFile()` is running the second SIGHUP may arrive and it would be ignored.